File: client/ios-objc/CardBrandChoice/CardBrandChoiceViewController.m
//
//  CheckoutViewController.m
//  app
//
//  Created by Ben Guo on 9/29/19, adapted by Nina Becx & Kevin Yang on 5/20/22.
//  Copyright © 2022 stripe-samples. All rights reserved.
//

#import "CardBrandChoiceViewController.h"
#import "DownPicker.h"
@import Stripe;

/**
* To run this app, you'll need to first run the sample server locally.
* Follow the "How to run locally" instructions in the root directory's README.md to get started.
* Once you've started the server, open http://localhost:4242 in your browser to check that the
* server is running locally.
* After verifying the sample server is running locally, build and run the app using the iOS simulator.
*/
NSString *const BackendUrl = @"http://127.0.0.1:4242/";

@interface CardBrandChoiceViewController ()

@property (nonatomic, weak) STPPaymentCardTextField *cardTextField;
@property (nonatomic, weak) UIButton *payButton;
@property (nonatomic, copy) NSString *paymentIntentClientSecret;
@property (nonatomic, weak) UITextField *dropdownTextField;
@property (strong, nonatomic) DownPicker *downPicker;

@end

@implementation CardBrandChoiceViewController

// Construct UI
- (void)viewDidLoad {
    [super viewDidLoad];

    self.view.backgroundColor = [UIColor whiteColor];

    // Card Form
    STPPaymentCardTextField *cardTextField = [[STPPaymentCardTextField alloc] init];
    self.cardTextField = cardTextField;
    [self.cardTextField addTarget:self action:@selector(onCardNumberChange) forControlEvents:UIControlEventValueChanged];

    // Dropdown
    UITextField *dropdownTextField = [[UITextField alloc] initWithFrame:CGRectMake(10, 200, 300, 40)];
    dropdownTextField.borderStyle = UITextBorderStyleRoundedRect;
    dropdownTextField.font = [UIFont systemFontOfSize:15];
    dropdownTextField.autocorrectionType = UITextAutocorrectionTypeNo;
    dropdownTextField.keyboardType = UIKeyboardTypeDefault;
    dropdownTextField.returnKeyType = UIReturnKeyDone;
    dropdownTextField.clearButtonMode = UITextFieldViewModeWhileEditing;
    dropdownTextField.contentVerticalAlignment = UIControlContentVerticalAlignmentCenter;
    self.dropdownTextField = dropdownTextField;

    self.downPicker = [[DownPicker alloc] initWithTextField:dropdownTextField withData:[[self brandMap] allKeys]];
    [self.downPicker setPlaceholder:@"Please Select a Card Brand"];

    // Pay Button
    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    button.layer.cornerRadius = 5;
    button.backgroundColor = [UIColor systemBlueColor];
    button.titleLabel.font = [UIFont systemFontOfSize:22];
    [button setTitle:@"Pay" forState:UIControlStateNormal];
    [button addTarget:self action:@selector(pay) forControlEvents:UIControlEventTouchUpInside];
    self.payButton = button;

    // Render all components
    UIStackView *stackView = [[UIStackView alloc] initWithArrangedSubviews:@[cardTextField, dropdownTextField, button]];
    stackView.axis = UILayoutConstraintAxisVertical;
    stackView.translatesAutoresizingMaskIntoConstraints = NO;
    stackView.spacing = 20;
    [self.view addSubview:stackView];

    [NSLayoutConstraint activateConstraints:@[
        [stackView.leftAnchor constraintEqualToSystemSpacingAfterAnchor:self.view.leftAnchor multiplier:2],
        [self.view.rightAnchor constraintEqualToSystemSpacingAfterAnchor:stackView.rightAnchor multiplier:2],
        [stackView.topAnchor constraintEqualToSystemSpacingBelowAnchor:self.view.topAnchor multiplier:2],
    ]];

    [self startCheckout];
}

// Mapping of brand external id to internal id
- (NSDictionary<NSString *, NSString *> *)brandMap {
    NSDictionary<NSString *, NSString *> *map = @{
        @"Visa" : @"visa",
        @"Mastercard" : @"mastercard",
        @"American Express" : @"amex",
        @"Discover" : @"discover",
        @"Diners Club" : @"diners",
        @"JCB" : @"jcb",
        @"UnionPay" : @"unionpay",
        @"Cartes Bancaires" : @"cartes_bancaires",
    };

    return map;
}

// Handler for observing the Card Number input field
- (void)onCardNumberChange {
    STPCardBrand brand = [STPCardValidator brandForNumber:self.cardTextField.cardNumber];
    NSString *brandString = [STPCardBrandUtilities stringFromCardBrand:brand];

    if (brandString != nil) {
        self.dropdownTextField.text = brandString;
    }
}

- (void)displayAlertWithTitle:(NSString *)title message:(NSString *)message restartDemo:(BOOL)restartDemo {
    dispatch_async(dispatch_get_main_queue(), ^{
        UIAlertController *alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
        if (restartDemo) {
            [alert addAction:[UIAlertAction actionWithTitle:@"Restart demo" style:UIAlertActionStyleCancel handler:^(UIAlertAction *action) {
                self.dropdownTextField.text = nil;
                [self.cardTextField clear];

                [self startCheckout];
            }]];
        }
        else {
            [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
        }
        [self presentViewController:alert animated:YES completion:nil];
    });
}

- (void)fetchPublishableKey {
    // Fetch publishable key by calling the sample server's /config endpoint.
    NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"%@config", BackendUrl]];
    NSMutableURLRequest *request = [[NSURLRequest requestWithURL:url] mutableCopy];
    [request setHTTPMethod:@"GET"];
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];

    NSURLSessionTask *task = [[NSURLSession sharedSession] dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *requestError) {
        NSError *error = requestError;
        NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
        if (error != nil || httpResponse.statusCode != 200 || json[@"publishableKey"] == nil) {
            dispatch_async(dispatch_get_main_queue(), ^{
                NSString *message = error.localizedDescription ?: @"";
                UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error loading page" message:message preferredStyle:UIAlertControllerStyleAlert];
                [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
                [self presentViewController:alert animated:YES completion:nil];
            });
        }
        else {
            NSString *stripePublishableKey = json[@"publishableKey"];
            // Configure the SDK with your Stripe publishable key so that it can make requests to the Stripe API
            [StripeAPI setDefaultPublishableKey:stripePublishableKey];
        }
    }];
    [task resume];
}

- (void)startCheckout {
    [self fetchPublishableKey];
    // Create a PaymentIntent by calling the sample server's /create-payment-intent endpoint.
    NSURL *url = [NSURL URLWithString:[NSString stringWithFormat:@"%@create-payment-intent", BackendUrl]];
    NSMutableURLRequest *request = [[NSURLRequest requestWithURL:url] mutableCopy];
    [request setHTTPMethod:@"POST"];
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];

    NSDictionary *jsonBodyDict = @{};
    NSData *jsonBodyData = [NSJSONSerialization dataWithJSONObject:jsonBodyDict options:kNilOptions error:nil];

    [request setHTTPBody:jsonBodyData];

    NSURLSessionTask *task = [[NSURLSession sharedSession] dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *requestError) {
        NSError *error = requestError;
        NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
        NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
        if (error != nil || httpResponse.statusCode != 200 || json[@"clientSecret"] == nil) {
            dispatch_async(dispatch_get_main_queue(), ^{
                NSString *message = error.localizedDescription ?: @"";
                UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error loading page" message:message preferredStyle:UIAlertControllerStyleAlert];
                [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
                [self presentViewController:alert animated:YES completion:nil];
            });
        }
        else {
            self.paymentIntentClientSecret = json[@"clientSecret"];
        }
    }];
    [task resume];
}

- (void)pay {
    if (!self.paymentIntentClientSecret) {
        NSLog(@"PaymentIntent hasn't been created");
        return;
    }

    [self.payButton setTitle:@"Paying..." forState:UIControlStateNormal];
    self.payButton.enabled = NO;
    
    // Set customer details
    STPPaymentMethodBillingDetails *billingDetails = [[STPPaymentMethodBillingDetails alloc] init];
    billingDetails.name = @"Jenny Rosen";

    // Collect card details on the client
    STPPaymentMethodCardParams *cardParams = self.cardTextField.cardParams;
    STPPaymentMethodParams *paymentMethodParams = [STPPaymentMethodParams paramsWithCard:cardParams billingDetails:billingDetails metadata:nil];
    STPPaymentIntentParams *paymentIntentParams = [[STPPaymentIntentParams alloc] initWithClientSecret:self.paymentIntentClientSecret];

    [[STPAPIClient sharedClient] createPaymentMethodWithParams:paymentMethodParams completion:^(STPPaymentMethod *paymentMethod, NSError *handleActionError) {
        if (paymentMethod) {
            NSArray *availableNetworks = paymentMethod.card.networks.available;
            NSString *network = [self brandMap][self.dropdownTextField.text];
            if ([availableNetworks containsObject:network]) {
                STPConfirmCardOptions *cardOptions = [[STPConfirmCardOptions alloc] init];
                STPConfirmPaymentMethodOptions *paymentMethodOptions = [[STPConfirmPaymentMethodOptions alloc] init];
                cardOptions.network = network;
                paymentMethodOptions.cardOptions = cardOptions;
                paymentIntentParams.paymentMethodOptions = paymentMethodOptions;
            }
            paymentIntentParams.paymentMethodId = paymentMethod.stripeId;
            [self confirmPayment:paymentIntentParams];
            return;
        }
        // For demo purposes, we expose the descriptive error message from the Stripe API
        [self displayAlertWithTitle:@"Payment method creation failed" message:handleActionError.userInfo[@"com.stripe.lib:ErrorMessageKey"] ?: @"" restartDemo:NO];
        [self.payButton setTitle:@"Pay" forState:UIControlStateNormal];
        self.payButton.enabled = YES;
    }];
}

- (void)confirmPayment:(STPPaymentIntentParams *)paymentIntentParams {
    // Complete the payment
    STPPaymentHandler *paymentHandler = [STPPaymentHandler sharedHandler];
    [paymentHandler confirmPayment:paymentIntentParams withAuthenticationContext:self completion:^(STPPaymentHandlerActionStatus status, STPPaymentIntent *paymentIntent, NSError *handleActionError) {
        dispatch_async(dispatch_get_main_queue(), ^{
            switch (status) {
                case STPPaymentHandlerActionStatusFailed: {
                    // For demo purposes, we expose the descriptive error message from the Stripe API
                    [self displayAlertWithTitle:@"Payment failed" message:handleActionError.userInfo[@"com.stripe.lib:ErrorMessageKey"] ?: @"" restartDemo:NO];
                    break;
                }
                case STPPaymentHandlerActionStatusCanceled: {
                    [self displayAlertWithTitle:@"Payment canceled" message:handleActionError.localizedDescription ?: @"" restartDemo:NO];
                    break;
                }
                case STPPaymentHandlerActionStatusSucceeded: {
                    [self displayAlertWithTitle:@"Payment succeeded" message:paymentIntent.description restartDemo:YES];
                    break;
                }
                default:
                    break;
            }

            [self.payButton setTitle:@"Pay" forState:UIControlStateNormal];
            self.payButton.enabled = YES;
        });
    }];
}

# pragma mark STPAuthenticationContext
- (UIViewController *)authenticationPresentingViewController {
    return self;
}

@end

File: client/ios-swift/CardBrandChoice/CardBrandChoiceViewController.swift
//
//  CardBrandChoiceViewController.swift
//  app
//
//  Created by Yuki Tokuhiro on 9/25/19, Adapted by Leo Chen on 05/26/22.
//  Copyright © 2022 stripe-samples. All rights reserved.
//

import UIKit
import Stripe
import DropDown

/**
 * To run this app, you'll need to first run the sample server locally.
 * Follow the "How to run locally" instructions in the root directory's README.md to get started.
 * Once you've started the server, open http://localhost:4242 in your browser to check that the
 * server is running locally.
 * After verifying the sample server is running locally, build and run the app using the iOS simulator.
 */
let BackendUrl = "http://127.0.0.1:4242/"

class CardBrandChoiceViewController: UIViewController {
    var paymentIntentClientSecret: String?
    
    var cardBrand: String?
    
    lazy var cardHolderFullNameField: UITextField = {
        let cardHolderFullNameField = UITextField()
        cardHolderFullNameField.placeholder = "Full Name"
        cardHolderFullNameField.borderStyle = .roundedRect
        return cardHolderFullNameField
    }()
    
    lazy var cardTextField: STPPaymentCardTextField = {
        let cardTextField = STPPaymentCardTextField()
        cardTextField.addTarget(self, action: #selector(editingCardNumberTextFieldValueChanged), for: .valueChanged)
        
        return cardTextField
    }()
    
    lazy var cardBrandChoiceButton: UIButton = {
        let button = UIButton(type: .custom)
        button.layer.cornerRadius = 5
        button.backgroundColor = .clear
        button.layer.cornerRadius = 5
        button.layer.borderWidth = 2
        button.layer.borderColor = UIColor.lightGray.cgColor
        button.titleLabel?.font = UIFont.systemFont(ofSize: 15)
        button.setTitle("Select Card Brand", for: .normal)
        button.setTitleColor(UIColor.black, for: .normal)
        // Show card brand dropdown when click the button
        button.addTarget(self, action: #selector(cardBrandChoiceDropdown), for: .touchUpInside)
        button.contentHorizontalAlignment = .left
        button.contentEdgeInsets = UIEdgeInsets(top: 10, left: 10, bottom: 10, right: 0);
        return button
    }()

    lazy var payButton: UIButton = {
        let button = UIButton(type: .custom)
        button.layer.cornerRadius = 5
        button.backgroundColor = .systemBlue
        button.titleLabel?.font = UIFont.systemFont(ofSize: 22)
        button.setTitle("Pay", for: .normal)
        button.addTarget(self, action: #selector(pay), for: .touchUpInside)
        return button
    }()
    

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .white
        let stackView = UIStackView(arrangedSubviews: [cardHolderFullNameField, cardTextField, cardBrandChoiceButton, payButton])
        stackView.axis = .vertical
        stackView.spacing = 20
        stackView.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(stackView)
        NSLayoutConstraint.activate([
            stackView.leftAnchor.constraint(equalToSystemSpacingAfter: view.leftAnchor, multiplier: 2),
            view.rightAnchor.constraint(equalToSystemSpacingAfter: stackView.rightAnchor, multiplier: 2),
            stackView.topAnchor.constraint(equalToSystemSpacingBelow: view.topAnchor, multiplier: 2),
        ])
        startCheckout()
    }
    
    @IBAction func cardBrandChoiceDropdown(_ sender: UIButton) {
        let dropDown = DropDown()
        let userFacingCardBrandNames = ["American Express", "Diners Club", "UnionPay", "Cartes Bancaires", "Discover", "JCB", "Mastercard", "Visa"]
        dropDown.dataSource = userFacingCardBrandNames
        dropDown.anchorView = sender
        dropDown.bottomOffset = CGPoint(x: 0, y: sender.frame.size.height)
        dropDown.show()
        dropDown.selectionAction = { [weak self] (index: Int, item: String) in
          guard let _ = self else { return }
        sender.setTitle(item, for: .normal)
        self?.cardBrand = CardBrandChoiceUtilities.toInternalCardBrandString(userFacingCardBrandName: item)
        }
      }
    
    @IBAction func editingCardNumberTextFieldValueChanged(_ sender: STPPaymentCardTextField) {
        let cardNumber = sender.cardNumber
        
        let brand = STPCardValidator.brand(forNumber: cardNumber ?? "")
        
        if brand != .unknown {
            let internalCardBrandStr = CardBrandChoiceUtilities.toInternalCardBrandString(brand: brand)
            let userFacingCardBrandStr = STPCardBrandUtilities.stringFrom(brand)
            cardBrandChoiceButton.setTitle(userFacingCardBrandStr, for: .normal)
            self.cardBrand = internalCardBrandStr
        }
    }
    
    func displayAlert(title: String, message: String, restartDemo: Bool = false) {
        DispatchQueue.main.async {
            let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
            if restartDemo {
                alert.addAction(UIAlertAction(title: "Restart demo", style: .cancel) { _ in
                    self.cardTextField.clear()
                    self.cardHolderFullNameField.text = nil
                    self.cardBrand = nil
                    self.cardBrandChoiceButton.setTitle("Select Card Brand", for: .normal)
                    self.startCheckout()
                })
            }
            else {
                alert.addAction(UIAlertAction(title: "OK", style: .cancel))
            }
            self.present(alert, animated: true, completion: nil)
        }
    }

    func startCheckout() {
        setStripePublishableKey()
        setPaymentIntentClientSecret()
    }
    
    func setStripePublishableKey() {
        let url = URL(string: BackendUrl + "config")!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        let task = URLSession.shared.dataTask(with: request, completionHandler: { [weak self] (data, response, error) in
            guard let response = response as? HTTPURLResponse,
                response.statusCode == 200,
                let data = data,
                let json = try? JSONSerialization.jsonObject(with: data, options: []) as? [String : Any],
                let stripePublishableKey = json["publishableKey"] as? String else {
                    let message = error?.localizedDescription ?? "Failed to decode response from server."
                    self?.displayAlert(title: "Error loading page", message: message)
                    return
            }
            StripeAPI.defaultPublishableKey = stripePublishableKey
        })
        task.resume()
    }
    
    func setPaymentIntentClientSecret(){
        // Create a PaymentIntent by calling the sample server's /create-payment-intent endpoint.
        let url = URL(string: BackendUrl + "create-payment-intent")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        let task = URLSession.shared.dataTask(with: request, completionHandler: { [weak self] (data, response, error) in
            guard let response = response as? HTTPURLResponse,
                response.statusCode == 200,
                let data = data,
                let json = try? JSONSerialization.jsonObject(with: data, options: []) as? [String : Any],
                  let clientSecret = json["clientSecret"] as? String else {
                    let message = error?.localizedDescription ?? "Failed to decode response from server."
                    self?.displayAlert(title: "Error loading page", message: message)
                    return
            }
            self?.paymentIntentClientSecret = clientSecret
        })
        task.resume()
    }
    @objc
    func pay(sender: UIButton) {
        guard let paymentIntentClientSecret = paymentIntentClientSecret else {
            return;
        }
        
        sender.setTitle("Paying...", for: UIControl.State.normal)
        sender.isEnabled = false
        
        // Collect card details
        let cardParams = cardTextField.cardParams
        
    
        // Collect the customer's full name to know which customer the PaymentMethod belongs to.
        let billingDetails = STPPaymentMethodBillingDetails()
        billingDetails.name = cardHolderFullNameField.text
        
        // Create PaymentIntent confirm parameters with the above
        let paymentMethodParams = STPPaymentMethodParams(card: cardParams, billingDetails: billingDetails, metadata: nil)
        
        let paymentIntentParams = STPPaymentIntentParams(clientSecret: paymentIntentClientSecret)

        paymentIntentParams.paymentMethodParams = paymentMethodParams
        
        if cardBrand != nil {
            let confirmCardOptions = STPConfirmCardOptions()
            let confirmPaymentMethodOptions = STPConfirmPaymentMethodOptions()
            confirmCardOptions.network = cardBrand
            confirmPaymentMethodOptions.cardOptions = confirmCardOptions
            paymentIntentParams.paymentMethodOptions = confirmPaymentMethodOptions
        }
        
        // Complete the payment
        let paymentHandler = STPPaymentHandler.shared()
        paymentHandler.confirmPayment(paymentIntentParams, with: self) { status, paymentIntent, error in
            switch (status) {
            case .failed:
                self.displayAlert(title: "Payment failed", message: error?.debugDescription ?? "")
                break
            case .canceled:
                self.displayAlert(title: "Payment canceled", message: error?.localizedDescription ?? "")
                break
            case .succeeded:
                self.displayAlert(title: "Payment succeeded", message: paymentIntent?.description ?? "", restartDemo: true)
                break
            @unknown default:
                fatalError()
                break
            }
            sender.setTitle("Pay", for: .normal)
            sender.isEnabled = true
        }
    }
}

extension CardBrandChoiceViewController: STPAuthenticationContext {
    func authenticationPresentingViewController() -> UIViewController {
        return self
    }
}

